core/system/src/mpu/vmpu_armv7m.c

```diff
@@ -146,60 +58,61 @@ static const TMpuRegion* vmpu_fault_find_region(uint32_t fault_addr)
+    const MpuRegion *region;
 
     /* return ACL if available */
-    switch (fault_addr) {
-        case (uint32_t) &SCB->SCR:
-            return UVISOR_TACL_UWRITE | UVISOR_TACL_UREAD;
-        default:
-            /* translate fault_addr into its physical address if it is in the bit-banding region */
-            if (fault_addr >= VMPU_PERIPH_BITBAND_START && fault_addr <= VMPU_PERIPH_BITBAND_END) {
-                fault_addr = VMPU_PERIPH_BITBAND_ALIAS_TO_ADDR(fault_addr);
-            }
-            else if (fault_addr >= VMPU_SRAM_BITBAND_START && fault_addr <= VMPU_SRAM_BITBAND_END) {
-                fault_addr = VMPU_SRAM_BITBAND_ALIAS_TO_ADDR(fault_addr);
-            }
+    if (fault_addr == (uint32_t) &SCB->SCR) {
+        return UVISOR_TACL_UWRITE | UVISOR_TACL_UREAD;
+    }
 
-            /* search base box and active box ACLs */
-            region = vmpu_fault_find_region(fault_addr);
+    /* translate fault_addr into its physical address if it is in the bit-banding region */
+    if (fault_addr >= VMPU_PERIPH_BITBAND_START && fault_addr <= VMPU_PERIPH_BITBAND_END) {
+        fault_addr = VMPU_PERIPH_BITBAND_ALIAS_TO_ADDR(fault_addr);
+    }
+    else if (fault_addr >= VMPU_SRAM_BITBAND_START && fault_addr <= VMPU_SRAM_BITBAND_END) {
+        fault_addr = VMPU_SRAM_BITBAND_ALIAS_TO_ADDR(fault_addr);
+    }
 
-            /* ensure that data fits in selected region */
-            if((fault_addr+size)>region->end)
-                return 0;
+    /* search base box and active box ACLs */
+    if (!(region = vmpu_fault_find_region(fault_addr))) {
+        return 0;
+    }
 
-            return region ? region->acl : 0;
+    /* ensure that data fits in selected region */
+    if((fault_addr + size) > region->end) {
+        return 0;
     }
+
+    return region->acl;
 }
 
+static int vmpu_mem_push_page_acl_iterator(uint8_t mask, uint8_t index);
+
 static int vmpu_fault_recovery_mpu(uint32_t pc, uint32_t sp, uint32_t fault_addr, uint32_t fault_status)
 {
-    const TMpuRegion *region;
-
-    /* Initialize the round-robin slot pointer. */
-    static uint32_t mpu_slot = ARMv7M_MPU_RESERVED_REGIONS;
+    const MpuRegion *region;
+    uint8_t mask, index, page;
 
     /* no recovery possible if the MPU syndrome register is not valid */
     if (fault_status != 0x82) {
         return 0;
     }
 
-    /* find region for faulting address */
-    if ((region = vmpu_fault_find_region(fault_addr)) == NULL)
-        return 0;
+    if (page_allocator_get_active_mask_for_address(fault_addr, &mask, &index, &page) == UVISOR_ERROR_PAGE_OK)
+    {
+        /* Remember this fault. */
+        page_allocator_register_fault(page);
 
-    /* In a secure box the first available region is for the stack, so it must
-     * be excluded from the round-robin. */
-    if (g_active_box != 0 && mpu_slot == ARMv7M_MPU_RESERVED_REGIONS) {
-        mpu_slot++;
+        vmpu_mem_push_page_acl_iterator(mask, UVISOR_PAGE_MAP_COUNT * 4 - 1 - index);
```

Comment:
@niklas-arm - the "-index" - isn't that a double negation of [#234](https://github.com/ARMmbed/uvisor/pull/294/files#diff-d5e475afa26cdc172546bc52c5d788abR234) ?